### PR TITLE
planner: cleanup unnecessary warning in join elimination

### LINKED ISSUE #9590: unnecessary and annoying warnings printed when eliminating outer join
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

```
mysql> create table a(c1 int);
Query OK, 0 rows affected (0.01 sec)

mysql> create table b(c1 int primary key);
Query OK, 0 rows affected (0.01 sec)

mysql> explain select * from a left join b on a.c1 = b.c1;
```
in TiDB log, we can see an empty warning log:
```
2019/03/07 14:51:14.766 rule_join_elimination.go:94: [warning] <nil>
```

2. What did you expect to see?

Not printing this empty log. We need to check if err is not nil before calling `log.Warn(err)` in line 94:
```golang
 86 // check whether schema cols of join's parent plan are all from outer join table
 87 func (o *outerJoinEliminator) isParentColsAllFromOuterTable(outerPlan LogicalPlan, parentSchema *expression.Schema) bool {
 88     if parentSchema == nil {
 89         return false
 90     }
 91     for _, col := range parentSchema.Columns {
 92         columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
 93         if c, err := outerPlan.Schema().FindColumn(columnName); c == nil {
 94             log.Warn(err)
 95             return false
 96         }
 97     }
 98     return true
 99 }
```

3. What did you see instead?

An empty warning log.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
Release Version: v3.0.0-beta-158-g247777d87
Git Commit Hash: 247777d8705fff77773508c833a40f0c00309784
Git Branch: master
UTC Build Time: 2019-03-07 06:49:36
GoVersion: go version go1.11.1 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve? 

Fix #9590
### What is changed and how it works?
Return err to optimizer instead of using log.warn which may cause unnecessary and annoying warnings printed when eliminating outer join.

### Check List 

Tests 

 - No code

Code changes

 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

### CODE REVIEW

File: planner/core/rule_join_elimination.go [hunk 1]
```diff
@@ -69,59 +85,68 @@ func (o *outerJoinEliminator) extractInnerJoinKeys(join *LogicalJoin, innerChild
...
-		return false
+		return false, nil
 	}
 	for _, col := range aggCols {
 		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
-		if c, err := outerPlan.Schema().FindColumn(columnName); c == nil {
-			log.Warn(err)
-			return false
+		c, err := outerPlan.Schema().FindColumn(columnName)
+		if err != nil {
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: Can we merge this `if` with the following `if` by:
```go
if err != nil || c == nil {
	return false, err
}
```

File: planner/core/rule_join_elimination.go [hunk 2]
```diff
@@ -69,59 +85,68 @@ func (o *outerJoinEliminator) extractInnerJoinKeys(join *LogicalJoin, innerChild
...
-		return false
+		return false, nil
 	}
 	for _, col := range parentSchema.Columns {
 		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
-		if c, err := outerPlan.Schema().FindColumn(columnName); c == nil {
-			log.Warn(err)
-			return false
+		c, err := outerPlan.Schema().FindColumn(columnName)
+		if err != nil {
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: Ditto.

File: planner/core/rule_join_elimination.go [hunk 3]
```diff
@@ -29,35 +28,52 @@ type outerJoinEliminator struct {
...
+		return nil, err
+	}
+	if matched {
+		return outerPlan, nil
 	}
 	// outer join elimination without duplicate agnostic aggregate functions
-	if !o.isParentColsAllFromOuterTable(outerPlan, parentSchema) {
-		return p
+	matched, err = o.isParentColsAllFromOuterTable(outerPlan, parentSchema)
+	if err != nil {
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: It's better to return `p, err` in the judgment.
Then we can merge the two `if` by:
```go
if err != nil || !matched {
	return p, err
}
```

File: planner/core/rule_join_elimination.go [hunk 4]
```diff
@@ -29,35 +28,52 @@ type outerJoinEliminator struct {
...
+	}
+	if !matched {
+		return p, nil
 	}
 	innerJoinKeys := o.extractInnerJoinKeys(p, innerChildIdx)
-	if o.isInnerJoinKeysContainUniqueKey(innerPlan, innerJoinKeys) {
-		return outerPlan
+	contain, err := o.isInnerJoinKeysContainUniqueKey(innerPlan, innerJoinKeys)
+	if err != nil {
+		return nil, err
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: Return `p, err`.

File: planner/core/rule_join_elimination.go [hunk 5]
```diff
@@ -29,35 +28,52 @@ type outerJoinEliminator struct {
...
+	}
+	if contain {
+		return outerPlan, nil
+	}
+	contain, err = o.isInnerJoinKeysContainIndex(innerPlan, innerJoinKeys)
+	if err != nil {
+		return nil, err
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: Ditto.

File: planner/core/rule_join_elimination.go (hunk 1 above)
Comment (wjhuang2016) [reply planner/core/rule_join_elimination.go]: Yes, it is better.

File: planner/core/rule_join_elimination.go (hunk 3 above)
Comment (wjhuang2016) [reply planner/core/rule_join_elimination.go]: Yes, it's better.

File: planner/core/rule_join_elimination.go [hunk 6]
```diff
@@ -173,25 +183,28 @@ func (o *outerJoinEliminator) isDuplicateAgnosticAgg(p LogicalPlan) (_ bool, col
...
 	if !isJoin {
-		return p
+		return p, nil
 	}
 	return o.tryToEliminateOuterJoin(join, aggCols, parentSchema)
 }
 
 func (o *outerJoinEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
-	return o.doOptimize(p, nil, nil), nil
+	return o.doOptimize(p, nil, nil)
```
Comment (lzmhhh123) [inline planner/core/rule_join_elimination.go]: It looks like we don't need `doOptimize` function. Could you pls remove it?

Comment (wjhuang2016) [reply planner/core/rule_join_elimination.go]: Sure, I will remove it.

Comment (wjhuang2016) [reply planner/core/rule_join_elimination.go]: It seems that we need the second and the thrid parameters.  Reomve it maybe not an good idea.

Comment (lzmhhh123) [reply planner/core/rule_join_elimination.go]: Yeah, you are right.

**Comment by lzmhhh123** [comment]
LGTM

**Pull request merged**